### ISSUE TITLE
Incorrect container CPU Usage reported with 4.7 kernel

### ISSUE THREAD

**Issue by pgeorgiev333** [moby/moby#26591]
**Description**

After updating my kernel to 4.7.3-100.fc23 on my fedora 23 host docker starts reporting strange container cpu usage.

**Steps to reproduce the issue:**
1. Upgrade the kernel to 4.7.3
2. Restart the host (this is a bare metal host)
3. Start docker engine and some containers that perform CPU intensive tasks. The containers have cpu_quota=500000 (with cpu_period=100000), i.e. should be able to use 5 cpus

**Describe the results you received:**
1. docker stats shows that the containers use ~2000% cpu (instead of around 500%)
2. I use cAdvisor with prometheus to monitor the various stats of the containers. With the 4.7 kernel the cAdvisor prometheus metric "container_cpu_usage_seconds_total" has entries for 100s of cpus, i,e, 

container_cpu_usage_seconds_total(cpu="00")
.....
container_cpu_usage_seconds_total(cpu="111")
container_cpu_usage_seconds_total(cpu="112")
..

The above messes up the prometheus reporting.

**Describe the results you expected:**
1. docker stats displays 400%-500% cpu usage (as the container's cpu usage is limited by the cpu_quota)
2. The cAdvisor prometheus metrics are correct:

http://<ip>:7080/metrics displays:

container_cpu_usage_seconds_total{cpu="cpu00"...)
...
container_cpu_usage_seconds_total{cpu="cpu39"...)

i.e. exactly 40 CPUS

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 21:39:21 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 21:39:21 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 16
 Running: 16
 Paused: 0
 Stopped: 0
Images: 5
Server Version: 1.11.2
Storage Driver: devicemapper
 Pool Name: docker-thinpool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 37.12 GB
 Data Space Total: 2.279 TB
 Data Space Available: 2.242 TB
 Metadata Space Used: 5.779 MB
 Metadata Space Total: 16.98 GB
 Metadata Space Available: 16.97 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.109 (2015-09-22)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local convoy
 Network: null host bridge
Kernel Version: 4.2.3-300.fc23.x86_64
Operating System: Fedora 23 (Twenty Three)
OSType: linux
Architecture: x86_64
CPUs: 40
Total Memory: 251.8 GiB
Name: docker-longevity.example.com
ID: ZTPL:EC67:NTBY:NZT5:AJFV:DKWO:GSJU:53V7:I5SD:IMKY:CI4Z:6Q4L
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
```

Note that in the above info i use 4.2.3 kernel where everything is working as expected.

Initially i thought that maybe the CPU_QUOTA was not being applied correctly and that the cpu usage of 2000% was ok (that should be 20 CPUS). But then i looked at the cAdvisor metrics and when i saw the huge number of container_cpu_usage_seconds_total() entries showing entries like cpu="111" i think that the issue may be purely reporting based.

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Physical environment (bare metal appliance).

Let me know if additional data is required as currently i run the box with the 4.2.3 kernel as i'm running a lot of tests and accurate cpu usage reporting is vital for test run comparison.

Thanks

**Comment by d-grigorenko** [comment]
Is it a dupe of https://github.com/moby/moby/pull/31579 ?

**Comment by pgeorgiev333** [comment]
I think yes

**Comment by thaJeztah** [comment]
Ah, yes, this is resolved in docker 17.03.1 (thanks for spotting @d-grigorenko), which adds an `OnlineCPUs` property to the stats endpoint to compensate for the breaking change in the kernel.

> I use cAdvisor with prometheus to monitor the various stats of the containers. With the 4.7 kernel the cAdvisor prometheus metric "container_cpu_usage_seconds_total" has entries for 100s of cpus, i,e,
> Describe the results you expected: The cAdvisor prometheus metrics are correct:

cAdvisor does not use the docker API for these metrics; they obtain this information directly from the kernel (similar to what the docker daemon does); so if this is not fixed in the current version of cAdvisor, I suggest to open a bug report with them.

I'll close this issue, but feel free to continue the conversation 👍
